boxoffice/models/line_item.py

```diff
@@ -1,47 +1,212 @@
-from decimal import Decimal
-from datetime import datetime
+import itertools
+import decimal
+from collections import namedtuple
+from boxoffice.models import db, BaseMixin, Order, Item, Price, DiscountPolicy
+from coaster.utils import LabeledEnum
+from baseframe import __
 
-from flask import url_for
+__all__ = ['LineItem']
 
-from boxoffice.models import db, BaseMixin
-from boxoffice.models.category import Category
-from boxoffice.models.order import Order
 
-__all__ = ['LineItem']
+class LINE_ITEM_STATUS(LabeledEnum):
+    CONFIRMED = (0, __("Confirmed"))
+    CANCELLED = (1, __("Cancelled"))
 
 
 class LineItem(BaseMixin, db.Model):
     """
-    A LineItem is part of an invoice.
-    Each line item carries an item category, quantity, charged amount, tax rate & line total.
+    Note: Line Items MUST NOT be deleted.
+    They must only be cancelled.
     """
-    __tablename__ = 'lineitem'
-    order_id = db.Column(db.Integer, db.ForeignKey('order.id'), nullable=False)
-    order = db.relation(Order, primaryjoin=order_id == Order.id,
-        backref=db.backref('lineitems', cascade='all, delete-orphan'))
-
-    #: Item category of this LineItem
-    category_id = db.Column(db.Integer, db.ForeignKey('category.id'), nullable=False)
-    category = db.relationship(Category, primaryjoin=category_id == Category.id)
-
-    quantity = db.Column(db.Integer, default=1, nullable=False)
-
-    total = db.Column(db.Numeric(10, 2), nullable=True)
-
-    def update_total(self):
-        self.total = self.quantity * self.category.price_after_tax
-
-    def dict(self):
-        return self.__dict__
-
-    def as_json(self):
-        return {
-            'event': self.category.event.title,
-            'category_id': self.category_id,
-            'category_desc': self.category.title,
-            'quantity': self.quantity,
-            'pat': self.category.price_after_tax,
-            'tax_rate': self.category.tax_rate,
-            'created_at': self.created_at.isoformat() + 'Z',
-            'updated_at': self.updated_at.isoformat() + 'Z',
-        }
+    __tablename__ = 'line_item'
+    __uuid_primary_key__ = True
+    __tableargs__ = (db.UniqueConstraint('order_id', 'line_item_no'),)
```

Comment:
The migrations weren't including the constraints, because of the typo. I've fixed the migration.